File: HW3_GrahamAlgorithm_CPP/GrahamScanner.cpp
#include <algorithm>
#include <cmath>
#include <string>
#include <vector>

#include "GrahamScanner.h"
#include "Stack.h"
#include "Point.h"

class GrahamScanner {
public:
	enum class OutputFormat {
		Plain, WKT
	};
	enum class Direction {
		Counterclockwise, Clockwise
	};

	~GrahamScanner() {
		delete convex_hull_;
	}

	GrahamScanner& CalculateConvexHull(std::vector<Point> points) {
		convex_hull_ = new Stack<Point>(points.size());
		if (points.empty()) {
			return *this;
		}
		// Saving the input.
		initial_input_ = points;

		Point start = FindStartingPoint(points);
		convex_hull_->Push(start);
		if (points.size() == 1) {
			return *this;
		}

		// Sorting the points by polar angle.
		std::sort(points.begin(), points.end(),
			[start](const Point& x, const Point& y) {
				return CompareByPolarThenDistance(x, y, start);
			});

		// The actual Graham algorithm.
		convex_hull_->Push(points[1]);
		for (size_t i = 2; i < points.size(); ++i) {
			PopWhileRightTurn(points[i]);
			convex_hull_->Push(points[i]);
		}
		// Connecting with the first point.
		PopWhileRightTurn(points[0]);

		return *this;
	}

	std::string GetData(GrahamScanner::Direction direction, OutputFormat format) {
		std::vector<Point> hull_vector = convex_hull_->ToVector();
		if (direction == Direction::Counterclockwise) {
			std::reverse(hull_vector.begin(), hull_vector.end());
		} else {
			// In this case we have right order initially,
			// but the start element is in the end.
			std::rotate(hull_vector.rbegin(), hull_vector.rbegin() + 1, hull_vector.rend());
		}

		if (format == OutputFormat::Plain) {
			return GetPlainFormat(hull_vector);
		} else if (format == OutputFormat::WKT) {
			return GetWKTFormat(hull_vector);
		}

		return "";
	}

private:
	Stack<Point>* convex_hull_{};
	std::vector<Point> initial_input_;

	static bool CompareByPolarThenDistance(const Point& point1, const Point& point2, const Point& start) {
		double p1_cos = -CalculatePolarCosine(start, point1);
		double p2_cos = -CalculatePolarCosine(start, point2);
		if (std::abs(p1_cos - p2_cos) < 0.00001) {
			return SquareDistance(start, point1) < SquareDistance(start, point2);
		} else {
			return p1_cos < p2_cos;
		}
	}

	static double CalculatePolarCosine(const Point& start, const Point& point) {
		if (point.x == start.x) {
			return point.y == start.y ? 2 : 0;
		}

		double hyp = std::sqrt(SquareDistance(start, point));
		double cat = static_cast<double>(point.x) - start.x;
		return cat / hyp;
	}

	Point FindStartingPoint(const std::vector<Point>& points) {
		Point start = points[0];
		for (const auto p : points) {
			if (p.y < start.y ||
				(p.y == start.y && p.x < start.x)) {
				start = p;
			}
		}
		return start;
	}

	std::string GetPlainFormat(const std::vector<Point>& hull) {
		std::string result = std::to_string(hull.size()) + '\n';
		if (hull.empty()) {
			return result;
		}
		for (size_t i = 0; i < (hull.size() - 1); ++i) {
			result += hull[i].toString();
			result += '\n';
		}
		result += hull.back().toString();
		return result;
	}

	std::string GetWKTFormat(const std::vector<Point>& hull) {
		std::string result = "MULTIPOINT((";
		if (!initial_input_.empty()) {
			for (size_t i = 0; i < (initial_input_.size() - 1); ++i) {
				result += initial_input_[i].toString();
				result += "), (";
			}
			result += initial_input_.back().toString();
		}
		result += "))\nPOLYGON ((";
		if (!hull.empty()) {
			for (const auto& p : hull) {
				result += p.toString();
				result += ", ";
			}
			result += hull[0].toString();
		}
		result += "))";

		return result;
	}

	bool IsRightTurn(const Point& point) {
		return ((convex_hull_->Top().x - convex_hull_->NextToTop().x) *
			(point.y - convex_hull_->Top().y) -
			(convex_hull_->Top().y - convex_hull_->NextToTop().y) *
			(point.x - convex_hull_->Top().x)) <= 0;
	}

	void PopWhileRightTurn(const Point& point) {
		while (convex_hull_->Size() > 1 && IsRightTurn(point)) {
			convex_hull_->Pop();
		}
	}
};

File: HW2_GrahamAlgorithm/GrahamScanner.cs
using System;
using System.Linq;
using System.Text;

namespace GrahamAlgorithm
{
    public class GrahamScanner
    {
        private Stack<Point> convexHull;
        private Point[] initialInput;

        /// <summary>
        /// Calculates the convex hull. To get result use <see cref="GetData(Direction, OutputFormat)"/>.
        /// </summary>
        public GrahamScanner CalculateConvexHull(Point[] points)
        {
            convexHull = new Stack<Point>(points.Length);
            if (points.Length == 0)
                return this;

            SaveInput(points);

            Point start = FindStartingPoint(points);
            convexHull.Push(start);
            if (points.Length == 1)
                return this;

            points = SortByPolarAngle(points, start);

            // The actual Graham algorithm.
            convexHull.Push(points[1]);
            for (int i = 2; i < points.Length; ++i)
            {
                PopWhileRightTurn(points[i]);
                convexHull.Push(points[i]);
            }
            // Connecting with the first point.
            PopWhileRightTurn(points[0]);

            return this;
        }

        /// <summary>
        /// Returns data as a string in a given format. 
        /// Throws error if the hull hasn't been calculated yet.
        /// </summary>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ArgumentException"></exception>
        public string GetData(Direction direction, OutputFormat format)
        {
            if (convexHull == null)
                throw new InvalidOperationException("Convex hull must be calculated first!");

            Point[] hullArray = convexHull.ToArray();
            hullArray = ChangeArrayDueToDirection(hullArray, direction);

            switch (format)
            {
                case OutputFormat.Plain:
                    return GetPlainFormat(hullArray);
                case OutputFormat.WKT:
                    return GetWKTFormat(hullArray);
                default:
                    throw new ArgumentException($"Format {format} is not supported!");
            }
        }


        private static Point[] ChangeArrayDueToDirection(Point[] hullArray, Direction direction)
        {
            switch (direction)
            {
                case Direction.Counterclockwise:
                    Array.Reverse(hullArray);
                    break;
                case Direction.Clockwise:
                    // In this case we have right order initially, 
                    // but the start element is in the end.
                    Point[] move = new Point[hullArray.Length];
                    for (int i = 1; i < hullArray.Length; ++i)
                    {
                        move[i] = hullArray[i - 1];
                    }
                    move[0] = hullArray[hullArray.Length - 1];
                    hullArray = move;
                    break;
                default:
                    throw new ArgumentException($"Direction {direction} is not supported!");
            }

            return hullArray;
        }

        /// <summary>
        /// Finds the starting point, which is left-est bottom-est one.
        /// </summary>
        private static Point FindStartingPoint(Point[] points)
        {
            Point start = points[0];
            foreach (var p in points)
            {
                if (p.Y < start.Y || (p.Y == start.Y && p.X < start.X))
                {
                    start = p;
                }
            }

            return start;
        }

        /// <summary>
        /// Checks whether the given point is making a "turn to the right" in respect to 
        /// the direction which is made by the last 2 points in the hull.
        /// The check is performed using cross product of the vectors.
        /// </summary>
        /// <param name="point"></param>
        /// <returns></returns>
        private bool IsRightTurn(Point point)
        {
            return (convexHull.Top().X - convexHull.NextToTop().X) * (point.Y - convexHull.Top().Y) -
                   (convexHull.Top().Y - convexHull.NextToTop().Y) * (point.X - convexHull.Top().X) <= 0;
        }

        /// <summary>
        /// Returns convex hull data in Plain format.
        /// </summary>
        private string GetPlainFormat(Point[] points)
            => new StringBuilder()
               .AppendLine(points.Length.ToString())
               .AppendJoin(Environment.NewLine, points)
               .ToString();

        /// <summary>
        /// Returns convex hull data in WKT format.
        /// </summary>
        private string GetWKTFormat(Point[] points)
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("MULTIPOINT ((")
              .AppendJoin("), (", initialInput)
              .AppendLine("))");

            sb.Append("POLYGON ((")
              .AppendJoin(", ", points)
              .Append($", {points[0]}))");

            return sb.ToString();
        }

        /// <summary>
        /// Pops top elements while the given point is making a "turn to the right" in respect to 
        /// the direction which is made by the last 2 points in the hull.
        /// </summary>
        private void PopWhileRightTurn(Point point)
        {
            while (convexHull.Size > 1 && IsRightTurn(point))
            {
                convexHull.Pop();
            }
        }

        private void SaveInput(Point[] points)
        {
            initialInput = new Point[points.Length];
            for (int i = 0; i < points.Length; ++i)
            {
                initialInput[i] = points[i];
            }
        }

        private Point[] SortByPolarAngle(Point[] points, Point start)
                    => points.OrderBy(x => Math.Round(-PolarCos(x, start), 6))
                             .ThenBy(x => Point.SquareDistance(start, x))
                             .ToArray();

        /// <summary>
        /// Calculates cosine of polar angle of <paramref name="p"/> relative to <paramref name="start"/>.
        /// </summary>
        private static double PolarCos(Point p, Point start)
        {
            // If directly above start (or even the start itself).
            if (p.X == start.X)
                return p.Y == start.Y ? 2 : 0;

            double hyp = Math.Sqrt(Point.SquareDistance(start, p));
            double cat = p.X - start.X;
            return cat / hyp;
        }


        public enum OutputFormat
        {
            Plain, WKT
        }

        public enum Direction
        {
            Counterclockwise, Clockwise
        }
    }
}
